sled-agent/src/storage/dump_setup.rs

```diff
@@ -0,0 +1,153 @@
+    }
+
+    fn run_dumpadm_and_savecore(
+        log: Logger,
+        dump_slices: Vec<Utf8PathBuf>,
+        u2_dump_dirs: Vec<Utf8PathBuf>,
+    ) {
+        for dump_slice in dump_slices {
+            // NOTE: because of the need to have dumpadm change the global
+            // state of which slice the system is using for dumps in order
+            // for savecore to behave the way we want (i.e. clear the flag
+            // after succeeding), we could hypothetically miss a dump if
+            // the kernel crashes again while savecore is still running.
+            if u2_dump_dirs.is_empty() {
+                // Don't risk overwriting an existing dump if there's
+                // already one there until we can attempt to savecore(8)
+                // it away and clear the flag to make room.
+                match illumos_utils::dumpadm::dump_flag_is_valid(&dump_slice) {
+                    Ok(false) => {
+                        // Have dumpadm write the config for crash dumps to be
+                        // on this slice, at least, until a U.2 comes along.
+                        match illumos_utils::dumpadm::dumpadm(&dump_slice, None)
+                        {
+                            Ok(_) => {
+                                info!(log, "Using dump device {dump_slice:?} with no savecore destination (no U.2 debug zvol yet)");
+                            }
+                            Err(err) => {
+                                warn!(log, "Could not configure {dump_slice:?} as dump device: {err:?}");
+                            }
+                        }
+                    }
+                    Ok(true) => {
+                        warn!(log, "Not configuring {dump_slice:?} as it appears to contain a dump we cannot yet send to a U.2 debug zvol");
+                    }
+                    Err(err) => {
+                        debug!(
+                            log,
+                            "Dump slice {dump_slice:?} appears to be unused : {err:?}",
+                        );
+                    }
+                }
+            } else {
+                // Try each U.2 until we succeed once
+                for mountpoint in &u2_dump_dirs {
+                    // Let's try to see if it appears to have a dump already
+                    match illumos_utils::dumpadm::dump_flag_is_valid(
+                        &dump_slice,
+                    ) {
+                        Ok(true) => {
+                            debug!(log, "Dump slice {dump_slice:?} appears to have a valid header; will attempt to savecore to {mountpoint:?}");
+                        }
+                        Ok(false) => {
+                            info!(log, "Dump slice {dump_slice:?} appears to have already been saved");
+                        }
+                        Err(err) => {
+                            debug!(log, "Dump slice {dump_slice:?} appears to be unused: {err:?}");
+                        }
```

Comment:
Should we be breaking out early in these cases? Or is the goal to map "dump slice" to "dump dir" even if no dump exists?

Reply:
on the contrary, in these cases we definitely want to configure the system to use the dump slice, as there's nothing important there to worry about overwriting